go/project_structure/library.md

```diff
@@ -0,0 +1,113 @@
+# Library
+
+This page describes common code structure for our libraries.
+
+The example below is taken from [k8sclient] on 2020/04/16. Most of the files
+and directories are elided for simplification.
+
+```
+github.com/giantswarm/k8sclient/
+├── CHANGELOG.md
+├── README.md
+├── go.mod
+├── go.sum
+└── pkg
+    ├── k8sclient
+    │   ├── clients.go
+    │   ├── error.go
+    │   ├── setup.go
+    │   ├── spec.go
+    │   ├── spec_test.go
+    │   └── types.go
+    ├── k8sclienttest
+    │   └── clients.go
+    ├── k8scrdclient
+    │   ├── crd_client.go
+    │   ├── crd_client_test.go
+    │   ├── error.go
+    │   └── spec.go
+    ├── k8srestconfig
+    │   ├── error.go
+    │   └── rest_config.go
+    └── k8sversion
+        ├── error.go
+        ├── funcs.go
+        ├── funcs_test.go
+        └── types.go
+```
+
+Below there are a few packages discussed.
+
+## Root package
+
+Note that since we use Go modules we don't have any Go files in the root of the
+repository. This is because as we bump the major version of the module we need
+to use `/vN/` imports. We put everything under subpackages of `/pkg` to avoid
+some ugly import aliases like:
+
+```
+k8sclientv3 "github.com/giantswarm/k8sclient/v3"
```

Comment:
In order to make that work the `go.mod` file needs to define `module github.com/giantswarm/k8sclient/v3`, which would be good to document as well. See https://github.com/giantswarm/k8sclient/blob/master/go.mod#L1.

Reply:
> This is because as we bump the major version of the module we needto use `/vN/` imports.

Isn't that the same thing? I'm not sure where to put this information with the current flow of the document. Suggestions welcome.

Reply:
I forgot to get the `go.mod` file right a couple of times already. The document here does not appear to tell me that. I am not sure if it should tell me that. I just wanted to bring forth the point and it kind of fits when you already suggest to have fancy imports. Because then the question is how to make them actually work.

Reply:
Also what I figured is that `module github.com/giantswarm/k8sclient/v1` does not work because given `vN` then `N >= 2`.